docs/compiler.rst

```diff
@@ -0,0 +1,273 @@
+Guide to the Scenic Parser & Compiler
+=====================================
+
+Architecture & Terminology
+--------------------------
+
+.. figure:: /images/parser_architecture.png
+  :alt: Scenic Parser & Compiler Architecture
+  :figclass: align-center
+
+Scenic AST
+~~~~~~~~~~
+
+
+Scenic AST is an abstract syntax tree for representing Scenic programs.
+It is a superset of Python AST and includes nodes for Scenic-specific
+language constructs.
+
+``syntax/ast.py`` defines all Scenic-specific AST nodes. Note that all
+Scenic AST nodes must extend the ``AST`` class defined on the same file.
+
+AST nodes should include fields to store objects. To add fields, add a
+parameter to the initializer and define fields by assigning values to
+``self``.
+
+When adding fields, be sure to update the ``_fields`` and
+``__match_args__`` fields. ``__fields`` lists the name of the fields on
+the AST node and is used by the AST module to traverse the tree, fill
+the missing information, etc. ``__match_args__`` is used by the test
+suite to assert the structure of the AST node using Python's structural
+pattern matching.
+
+Scenic Grammar
+~~~~~~~~~~~~~~
+
+Scenic Grammar (``scenic.gram``) is a formal grammar that defines the syntax
+of the Scenic language. It is written as a parsing expression grammar
+(PEG) using `the Pegen parser generator <https://we-like-parsers.github.io/pegen/index.html>`_.
+
+Please refer to `Pegen's documentation <https://we-like-parsers.github.io/pegen/grammar.html>`_ on how to write a grammar.
+
+Scenic Parser
+~~~~~~~~~~~~~
+
+Scenic Parser takes a Scenic source code and outputs the corresponding
+abstract syntax tree. It is generated from the grammar file using Pegen.
+
+When you modify ``scenic.gram``, you need to generate the parser again.
+Call ``make`` at the project root to do so. Pytest is configured to
+update the parser before test execution.
+
```

Comment:
I believe that the makefile won't work for Windows users outside of WSL, so maybe we should add a note here on how to run the command manually for them?

Reply:
Also does Pytest update the parser before test execution (if changes have been made), or just generate it if it doesn't already exist?